Adding maintainers to opensearch-project. (#59)
opensearch-project/.github: MAINTAINERS.md

  - [Triage Open Issues](#triage-open-issues)
  - [Be Responsive](#be-responsive)
  - [Maintain Overall Health of the Repo](#maintain-overall-health-of-the-repo)
  - [Add Continuous Integration Checks](#add-continuous-integration-checks)
    - [Developer Certificate of Origin Workflow](#developer-certificate-of-origin-workflow)
  - [Use Semver](#use-semver)
  - [Release Frequently](#release-frequently)
  - [Promote Other Maintainers](#promote-other-maintainers)
  - [Describe the Repo](#describe-the-repo)
- [Becoming a Maintainer](#becoming-a-maintainer)
  - [Nomination](#nomination)
  - [Interest](#interest)
  - [Public Confirmation](#public-confirmation)
  - [Maintainer Decision](#maintainer-decision)

>   - [Adding the New Maintainer](#adding-the-new-maintainer)

Reviewer: Nit : It should be `Adding a New Maintainer` right?

Author: Corrected.

- [Removing a Maintainer](#removing-a-maintainer)
  - [Moving On](#moving-on)
  - [Inactivity](#inactivity)
  - [Negative Impact on the Project](#negative-impact-on-the-project)

>    

Author: There are no new rules on how this document can be modified. It's an important, future question.

>    

Reviewer 2: @dblock  What I was getting at was how this works on the repo local level. So, this is a `.github` (org) level document. Usually those can be overridden at the repo level - is that still OK? If, say opensearch-cli wants to add extra or modify this, is that okay? 

I think we should just spell that out somewhere.

>    

Author: I see your point. Most MAINTAINERS.md in repos are actually redirects to this document, e.g. https://github.com/opensearch-project/anomaly-detection/blob/main/MAINTAINERS.md. Some are not. Either way it's a much bigger question, so I opened https://github.com/opensearch-project/.github/issues/60.

### Promote Other Maintainers

Assist, add, and remove [MAINTAINERS](MAINTAINERS.md). Exercise good judgement, and propose high quality contributors to become co-maintainers. See [Becoming a Maintainer](#becoming-a-maintainer) for more information.

### Describe the Repo

Make sure the repo has a well-written, accurate, and complete description. See [opensearch-project/.github#38](https://github.com/opensearch-project/.github/issues/38) for some helpful tips to describe your repo.

## Becoming a Maintainer

You can become a maintainer by actively [contributing](CONTRIBUTING.md) to any project, and being nominated by an existing maintainer.

### Nomination

> Any current maintainer starts a private e-mail thread (until we have a better mechanism, e-mail addresses can usually be found via MAINTAINERS.md + DCO) with all other maintainers on that repository to discuss nomination using the template below. In order to be approved, at least three positive maintainer votes are necessary, and no vetoes. In rare cases when there are fewer than three maintainers, the positive votes from all maintainers are required. Any disagreements can be escalated to the repo admin.

Reviewer 3: It would be great if we can automate the nomination process based on people's contribution statistics and void depending on maintainers to initiate the process

Author: It's a good idea. I don't think we need to be spelling anything about it here though. Let me know if you strongly disagree.

The nomination should clearly identify the person with their real name and a link to their GitHub profile, and the rationale for the nomination, with concrete example contributions.

### Interest

> Upon receiving at least three positive maintainer votes, and no vetoes, from other maintainers within a maximum period of two weeks, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository via private e-mail message.

Reviewer 4: I would change the period to one week.  Maintainers in general show strong ownership, checking emails often for high priority things like VOTEing a new maintainer, and the "at least three" protects against too many maintainers going on no-internet vacations at once.

I propose re-wording this to "Upon receiving three or more +1 maintainer votes, and no -1 vetoes, from existing maintainers, after a one week vote, the nominating maintainer....".  Using `maximum` is too ambiguous here -- it could mean that after one hour if three maintainers +1 then the VOTE is done.

Reviewer 5: +1 for the change, but fine if this happens in a follow up PR. 

```suggestion
Upon receiving at least three positive (+1) maintainer votes, and no vetoes (-1), from existing maintainers after a one week period, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository via private e-mail message.
```

> Upon receiving at least three positive maintainer votes, and no vetoes, from other maintainers within a maximum period of two weeks, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository, which happens via private e-mail message, but may also be a comment to a pull request, following a substantial contribution.

Reviewer 4: s/asks a potential nominee/asks the potential nominee

`within a maximum period of two weeks` is odd -- did you mean minimum period maybe?   Or maybe "at least one week and at most two weeks"?

If it is meant to be minimum, two weeks is too long imo.

I still think we should remove the `, but may also be ...` clause -- asking the nominee should be a private email only -- people decline for all sorts of important personal reasons, and that part should remain private as well.

Author: It's meant to say that maintainers will wait for max 2 weeks to collect votes and vetoes. Hopefully maintainers are active enough to respond quicker than that. So it's not a minimum, nor is it at least one week and at most two. I don't know how to wordsmith this better.

Author: I don't care enough to allow asking in public, removed. 

https://github.com/opensearch-project/.github/pull/59/commits/732388472c9f100465bfa3fac33e73a8e35f99ec

> Upon receiving at least three positive maintainer votes, and no vetoes, from other maintainers within a maximum period of two weeks, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository, which happens via private e-mail message message, but may also be a comment to a pull request, following a substantial contribution.

Reviewer 6: ```suggestion
Upon receiving at least three positive maintainer votes, and no vetoes, from other maintainers within a maximum period of two weeks, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository, which happens via private e-mail message, but may also be a comment to a pull request, following a substantial contribution.
```

Author: https://github.com/opensearch-project/.github/pull/59/commits/6a1fc434e0df421f864b7a30d7887d2a51077798

> > This is great work! Based on your ongoing engagement with the project, the current maintainers now invite you to become a co-maintainer for this project. Please respond and let us know if you would like to become maintainer.

Reviewer 5: ```suggestion
> This is great work! Based on your valuable contribution and ongoing engagement with the project, the current maintainers invite you to become a co-maintainer for the OpenSearch project. Please respond and let us know if you accept the invitation to become maintainer.
```

Author: https://github.com/opensearch-project/.github/pull/59/commits/9580d11a4f1bbf39ade9cf0ba8d66bed2b40ffb7

I didn't keep the "the OpenSearch project" because this is generic for all repos and I don't want to misrepresent that the maintainer is only being added to a specific repo. I don't want to call the repo a repo though, we call them projects.

Individuals accept the nomination by replying, or commenting, for example _"Thank you! I would love to."_

### Addition

> Upon receiving three positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. The pull request is approved and merged.

Reviewer 4: add `privately` after `... confirmed interest with the nominee`

Author: https://github.com/opensearch-project/.github/pull/59/commits/732388472c9f100465bfa3fac33e73a8e35f99ec

### Promote Other Maintainers

Assist, add, and remove [MAINTAINERS](MAINTAINERS.md). Exercise good judgement, and propose high quality contributors to become co-maintainers. See [Becoming a Maintainer](#becoming-a-maintainer) for more information.

### Describe the Repo

Make sure the repo has a well-written, accurate, and complete description. See [opensearch-project/.github#38](https://github.com/opensearch-project/.github/issues/38) for some helpful tips to describe your repo.

## Becoming a Maintainer

You can become a maintainer by actively [contributing](CONTRIBUTING.md) to any project, and being nominated by an existing maintainer.

### Nomination

> Any current maintainer starts a private e-mail thread (until we have a better mechanism, e-mails can usually be found via MAINTAINERS.md + DCO) with all other maintainers on that repository to discuss nomination using the template below. In order to be approved, at least 3 positive maintainer votes are necessary, and no vetoes. In rare cases when there are fewer than 3 maintainers, the positive votes from all maintainers are required. Any disagreements can be escalated to the repo admin.

Reviewer 4: s/`e-mails`/`e-mail addresses`?

Reviewer 4: If possible, please line-break these long paragraphs so that we can comment on specific lines here in GitHub :)

Author: https://github.com/opensearch-project/.github/pull/59/commits/0851f1de871f6457e4df65285355d49b7805707d

The nomination should clearly identify the person with their real name and a link to their GitHub profile, and the rationale for the nomination, with concrete example contributions.

### Interest

> Upon receiving at least 3 positive maintainer votes, and no vetoes, from other maintainers within a maximum period of 2 weeks, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository, which typically happens via e-mail obtained via DCO, but may also be a comment to a pull request, following a substantial contribution.

Reviewer 4: s/3/three
s/2/two

Remove "typically" and make it more concrete: `, which happens via private e-mail message`, and remove the `, but may also be a ...`?

Author: Why would it be a problem to ask someone to join as a maintainer in public?

Reviewer 7: While the primary issue is the public *discussion* of the merits of an individual, not the public invitation itself, I would offer one perspective on why a public invitation can be harmful. It creates a situation where now everyone knows that they've been invited, and there will (could?) be pressure from their employer to accept an invitation that they don't have the time or inclination to accept. A private invitation allows the individual to make the decision in their own time, in their own way, without external pressure or judgement. Granted, this may be a non-issue for many/most people, but I feel that it's the more sensitive approach.

Reviewer 4: > Why would it be a problem to ask someone to join as a maintainer in public? I do this all the time on my projects and people sometimes say they just don't have bandwidth for it.

I think this is in reference to me asking to remove the `, but...` clause right?

Multiple reasons:
  * Inviting a strong community contributor to become a possible new maintainer should not be seen by the community as an ad-hoc comment on a PR, but rather a well thought out discuss/VOTE, privately, among the existing committers (yes, in response to a strong code or community contribution, or earned merit from strong community engagement, but nevertheless a private VOTE among the existing maintainers).  It should be a new email, announcing the successful (private) vote outcome.
  * The subsequent discussion about whether the proposed new maintainer agrees that they would like to join the project or not, should also remain private.  There are many complex personal reasons about why an individual may or may not want to join, and it is their decision, not ours, whether that is a public discussion.  I have seen so many interesting (private, legitimately) reasons over the years.
  * Can you point to "my projects" so we can make an informed comparison?  I agree for smaller informal projects, such an informal process makes sense.  But this is not that.

Author: Yes, these are "smaller" or "informal", indeed. But maybe we would be better served to treat our modest beginnings as such, as well and not make too big of a deal of this 🤔 

https://github.com/slack-ruby/slack-ruby-client/pull/387#issuecomment-988398343
https://github.com/alexa-js/alexa-app/pull/310#pullrequestreview-92028980
https://github.com/mongoid/mongoid-slug/pull/252#issuecomment-413296153

> > This is great work! Based on your ongoing contributions to the project we have decided that we'd like to invite you to become a co-maintainer for this project. Please let us know whether you accept this nomination.

Reviewer 4: Make it more concrete: `Based on your ongoing engagement with the project, the current maintainers now invite you to become...`

Author: https://github.com/opensearch-project/.github/pull/59/commits/0851f1de871f6457e4df65285355d49b7805707d

> Individuals accept the nomination by replying, or commenting, for example _"Thank you! I would love to."_

Reviewer 4: The e-mail message template above should be fixed to make this explicit: `Please respond to this e-mail message letting us know if you would like to become maintainer....`

Author: https://github.com/opensearch-project/.github/pull/59/commits/6947878111ed664be9225f0a2410a2cfed4be468

### Public Confirmation

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Author: Do you have any proposed language?

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Author: It's a proposal and vote, updated the text.

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Author: https://github.com/opensearch-project/.github/pull/59/commits/f2e5328d5bb0e5201a040a67fc3e3e498eca0600

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Reviewer 8: I ask in part because I'm not sure what the exit requirements are here.

But, I'll give is a stab based on what I've gathered. If I understand what the overall process correctly, I'd modify the "Maintainer Decision" section as follows:

> At the end of two weeks, the maintainers will perform a final vote similar to the nomination process. If the nomination still has three positive votes, and no vetoes, from other maintainers, the nomination is accepted. The positive maintainer voters should approve the PR to show support. The PR may be merged with only one approver.

> If the maintainers cannot reach a positive decision for any reason, the nomination is rejected with "a decision could not be reached" with no further explanation necessary, and the pull request is closed.

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Reviewer 8: I ask because this doesn't seem like a vote to me. It seems like a public comments period. As I understand it, the maintainers are still making the decision and there is no "vote" to make visible to the community. There is only an opportunity for community members to raise concerns.

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Author: Maintainers still have to publicly re-add their +1 to ack their agreement, that's what I meant. Help me make that more clear?

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Author: Ok, I see, this is clearer. Check out https://github.com/opensearch-project/.github/pull/59/commits/624799078f9413b1fa0f5a0122c0e731cc338c23.

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md. This makes the vote visible to the community. 

Author: https://github.com/opensearch-project/.github/pull/59/commits/624799078f9413b1fa0f5a0122c0e731cc338c23

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a pull request adding the proposed co-maintainer to MAINTAINERS.md, and perform a final vote similar to the nomination process. If the nomination still has 3 positive votes, and no vetoes, from other maintainers, the nomination is accepted. The positive maintainer voters should approve the pull request to show support.

Reviewer 4: s/3/three
s/the maintainer/the nominating maintainer

This opening sentence should state the timeframe, e.g. `after 72 hours, upon receiving three positive maintainer votes and no vetoes...`, `and after reaching consent with the nominee", ...

The first notification to the community should NOT be a PR, but rather a celebratory email to the dev list announcing that maintainers asked contributor X to become a new maintainer and they/she/he agreed!  In Apache we usually ask the new committer/maintainer to then author and eventually push the PR, making it all official.

Author: I removed the public confirmation in https://github.com/opensearch-project/.github/pull/59/commits/6947878111ed664be9225f0a2410a2cfed4be468 and there's a 2 week period for internal parts. I think that addresses all these concerns, but re-add comments if otherwise.

https://github.com/opensearch-project/.github/pull/59/commits/6947878111ed664be9225f0a2410a2cfed4be468

> _Content from the above nomination._
> 
> Here is my +1.
> 
> If you support this nomination, please add your +1 or a comment. 
> 

> > If you do not support this nomination, we invite you to e-mail all/some of the current [project maintainers](https://github.com/opensearch-project/OpenSearch/blob/main/MAINTAINERS.md) in the next 2 weeks. We will keep your feedback private to this group and discuss offline. We will not publicly post the reasons for a decision not to add this co-maintainer, or publicize your name. 

Reviewer 4: -1 to all of this!

We trust the existing maintainers to have the best interests of the project in mind.  We should not have a public (and infinitely archived) email discussion about this person's merits.  Please, be inspired by or simply poach how Apache has handled this after ~23 years of lessons learned.

Reviewer 5: > we invite you to e-mail all/some of the current [project maintainers]

Agree. The only public feedback mechanism for this should be a private distro list the public can use to email all maintainers about inappropriate behavior or concerns in the public. The maintainers can then have private discussions about any CoC violations. If the inappropriate behavior involves a maintainer this is where one might lose their commit bit.

Author: https://github.com/opensearch-project/.github/pull/59/commits/6947878111ed664be9225f0a2410a2cfed4be468

> 
> Anyone, including maintainers, can publicly add +1s as requested in the nomination issue if they support the nomination or contact the maintainers in private if they oppose. The pull request stays open for 2 weeks.

### Maintainer Decision

If a positive decision has been reached, the nomination is accepted. If the maintainers cannot reach a positive decision for any reason, the nomination is rejected with _"a decision could not be reached"_ with no further explanation necessary, and the pull request is closed.

### Adding a New Maintainer

The repo admin adjusts the new maintainer’s permissions accordingly, and merges the pull request to MAINTAINERS.md.

## Removing a Maintainer

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Reviewer 5: Can we change this language so it doesn't read that inactivity will cause you to be removed as a maintainer? It may just suspend your access but you're still a maintainer. I think this is more consistent with "merit doesn't expire". 

```suggestion
Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as violating the [code of conduct](https://github.com/opensearch-project/.github/blob/main/CODE_OF_CONDUCT.md), or taking other actions that negatively impact the project.
```

Author: Agreed.

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as violating the [code of conduct](https://github.com/opensearch-project/.github/blob/main/CODE_OF_CONDUCT.md), or taking other actions that negatively impact the project.

Reviewer 9: We should be explicit here about actions that we already know will result in loss of maintainer status. Violating code of conduct is one, but the "might be" language can lead to trouble. We may not know everything, but we know some things

Reviewer 9: Do we need an explicit process for contributors and admins to start a removal?

Author: Agreed. I changed _might_ to _will_ in https://github.com/opensearch-project/.github/pull/59/commits/6d68d527c1e48033708f824843ecad4e27d51291.

### Moving On

There are plenty of reasons that might cause someone to want to take a step back or even a hiatus from a project. Existing maintainers can choose to leave the project at any time, with or without reason, by making a pull request to move themselves to the "Emeritus" section of MAINTAINERS.md, and asking an admin to remove their permissions.

### Inactivity

> If the maintainers on the project notice that another maintainer is no longer an active or visible member of the community, the first step will be to tell the repo admin to contact the inactive maintainer to check-in. If the inactive maintainer responds and reengages, no action is required. If the maintainer does not respond after a week, or indicates they no longer wish to be a maintainer, the repo admin can remove the inactive maintainer’s permissions and make a pull request to add them to the "Emeritus" section of the MAINTAINERS.md.

Author: We had an earlier comment talking about vacations in Europe being 4 weeks to several months!

I am going to change this to "several weeks" to avoid being very prescriptive.

> If the maintainers on the project notice that another maintainer is no longer an active or visible member of the community, the first step will be to tell the repo admin to contact the inactive maintainer to check-in. If the inactive maintainer responds and reengages, no action is required. If the maintainer does not respond after a week, or indicates they no longer wish to be a maintainer, the repo admin can remove the inactive maintainer’s permissions and make a pull request to add them to the "Emeritus" section of the MAINTAINERS.md.

Author: https://github.com/opensearch-project/.github/pull/59/commits/b5d083598d95016dc2e76c7f8475a00f1b4e24ae

> If the repo is left without any maintainers, either by maintainer inactivity or moving on, the repo is considered unmaintained. The repo admin will seek out new maintainers and note the maintenance status in the repo README file.

Reviewer 5: ```suggestion
If the repo is left without any maintainers, either by maintainer inactivity or moving on, the repo is considered unmaintained. The repo admin will seek out new maintainers and note the maintenance status in the repo README file. This could lead to a repo becoming archived due to lack of maintainers. 
```

Author: Archiving repos is a separate topic. I do like this, but I will keep it out of this PR.

Author: > But I don't think the wider community should have a chance to publicly discuss this individual's additional commit/maintainer karma -- it should only be up to the existing maintainers to make that decision (and then announce it widely, publicly).

Let's agree to disagree. There's nothing preventing public comment from happening, certainly not a delay of keeping the issue open for a couple of weeks. The text in the PR is inviting people to comment in private, and during that time someone reasonable may offer a reasonable argument to not add a maintainer, and e-mail the maintainers in private. If we just go ahead, they are more likely to manifest their disagreement in public, which I think is net worse.

All that said, if someone else feels strongly otherwise, I'd like to hear from you!

Reviewer 5: > Archiving repos is a separate topic.

Good point. Probably best to add this language after there's a process to put a project in maintenance mode. I've [opened an issue](https://github.com/opensearch-project/.github/issues/68) for this. The Linux Foundation [has some good material](https://www.linuxfoundation.org/tools/winding-down-an-open-source-project/) we can leverage. I'd also be curious to see what @ke4qqq @rbowen think

Reviewer 7: I do feel strongly about it for a couple of reasons.

One is historical bias (ie, "we've always done it that way") from Apache involvement. Admittedly, that's never a great reason.

But I think it goes deeper than that. Having discussion of personnel issues (and that's what this is) in the public record - particularly in a public mailing list, which will be archived, permanently, forever, can open us ("us" = project, company, individuals) up to legal problems, and can follow the person being discussed throughout the rest of their career. For this reason, at Apache at least, we have always held these kinds of conversations private.

The other aspect of this is that it can be deeply humiliating, and career-limiting, to have a public conversation (again, permanently archived) where one's name is suggested for commit/maintainer, and the conclusion is "no". Potential employers will find that conversation forever, and that is hugely unfair (not to mention the above legal consideration) to the person in question.

So, yes, for these reasons, and perhaps others, we should steer all discussion of proposed committers and maintainers to a private list where people are expected to respect confidentiality.

Author: https://github.com/opensearch-project/.github/pull/59/commits/6947878111ed664be9225f0a2410a2cfed4be468

### Negative Impact on the Project

> Actions that negatively impact the project will be handled by the admins, in coordination with other maintainers, in balance with the urgency of the issue. Examples would be [Code of Conduct](CODE_OF_CONDUCT.md) violations, deliberate harmful actions, and security risks.

Reviewer 5: ```suggestion
Actions that negatively impact the project will be handled by the admins, in coordination with other maintainers, in balance with the urgency of the issue. Examples would be [Code of Conduct](CODE_OF_CONDUCT.md) violations, deliberate harmful or malicious actions, and security risks.
```

> Actions that negatively impact the project will be handled by the admins, in coordination with other maintainers, in balance with the urgency of the issue. Examples would be [Code of Conduct](CODE_OF_CONDUCT.md) violations, deliberate harmful actions, and security risks.

Reviewer 8: Should we have some clarity for what maintainers should do if they believe that another maintainer is having a negative impact on the project? A repo admin may not be as aware of any such violations as other maintainers.

Author: We used to have prescriptive text about that a maintainer needs to email other maintainers and the admin, but it became very specific and not very effective in all cases. There were all these edges where it just wouldn't work, so I prefer to keep this general for now.

### Promote Other Maintainers

Assist, add, and remove [MAINTAINERS](MAINTAINERS.md). Exercise good judgement, and propose high quality contributors to become co-maintainers. See [Becoming a Maintainer](#becoming-a-maintainer) for more information.

### Describe the Repo

Make sure the repo has a well-written, accurate, and complete description. See [opensearch-project/.github#38](https://github.com/opensearch-project/.github/issues/38) for some helpful tips to describe your repo.

## Becoming a Maintainer

> You can become a maintainer by actively [contributing](CONTRIBUTING.md) to any project, and being nominated by an existing maintainer.

Reviewer 10: The criteria to nominate someone can vary greatly between maintainers. How do we make sure that the maintainers are not constantly blocking each other's nominees?

Reviewer 11: I think @nknize mention he'd seen something similar, and that people would basically just re-nominate (there's nothing to stop you from nominating someone again).  I'd be curious to hear other's thoughts on better guardrails.

Author: I think it's absolutely OK. There can always be a maintainer that vetoes everything all the time. There can always be someone nominating random people. Any group of people needs to find a way to productively disagree, and YMMV. 

In my experience, those who became maintainers have had an opportunity to earn trust in a way that doesn't prevent progress, so it won't happen often. And we have an escalation path to the admins for resolving this if it happens.

Reviewer 12: I agree with @dblock, the similar consensus rules are used by ASF and OpenJDK for example, and usually the process is driven by good will (whenever there are a -1/no votes, the compelling reasons are expected to be given, I saw that only few times in many years).

### Nomination

> Any current maintainer starts an e-mail thread (until we have a better mechanism, e-mails can usually be found via MAINTAINERS.md + DCO) with all other maintainers on that repository to discuss nomination using the template below. In order to be approved, at least 3 positive maintainer votes are necessary, and no vetoes. Any disagreements can be escalated to the repo [Admin](ADMINS.md).

Reviewer 10: So we need at least 3 maintainers per repo. I think if there are fewer then there should be maintainers from other repos taking part in the vote.

Reviewer 11: Hmmm....It may be that some repos aren't keeping their maintainer docs up to date.  :(  Which repos were you thinking of?

Reviewer 10: It is dynamic. As stated some maintainers can take a hiatus, or be suspended for some reason. So it is a possibility unless we say something like - "All repos require a minimum of 3 maintainers" ?

Author: While we do try to ensure a minimum of maintainers, I think we might end up with repos with < 3. I thought it was pretty obvious that you can't have 3 people vote when there are fewer than 3 people on the list and didn't want to write that, but open for a specific edit if you think that's worth it?

Reviewer 12: @dblock may be phrase it like that: "In order to be approved, at least 3 positive maintainer votes are necessary, and no vetoes. In rare cases when there are less than 3 maintainers, the positive votes from all maintainers are required.", wdyt?

Author: Added. I like it.

> Any current maintainer starts an e-mail thread (until we have a better mechanism, e-mails can usually be found via MAINTAINERS.md + DCO) with all other maintainers on that repository to discuss nomination using the template below. In order to be approved, at least 3 positive maintainer votes are necessary, and no vetoes. In rare cases when there are less than 3 maintainers, the positive votes from all maintainers are required. Any disagreements can be escalated to the repo admin.

Reviewer 4: Also, replace `less than` with `fewer than`?  Countable nouns ...

> Any current maintainer starts an e-mail thread (until we have a better mechanism, e-mails can usually be found via MAINTAINERS.md + DCO) with all other maintainers on that repository to discuss nomination using the template below. In order to be approved, at least 3 positive maintainer votes are necessary, and no vetoes. In rare cases when there are less than 3 maintainers, the positive votes from all maintainers are required. Any disagreements can be escalated to the repo admin.

Author: https://github.com/opensearch-project/.github/pull/59/commits/7c262a401718f46ccc813844d92a7cdca5a89f4c

> > I propose inviting Vacha Shah ([@vachashah](https://github.com/vachashah)) as a co-maintainer on [opensearch-project/OpenSearch](https://github.com/opensearch-project/OpenSearch). 

Author: Yeah, I did ask Vacha before using her as an example, but I prefer your suggestion. Changed in https://github.com/opensearch-project/.github/pull/59/commits/f4e8ba2804a06a88eb4fa137260de41e3ed8548a.

> 
> Vacha has contributed a lot of automation across many repositories.
> 
> - https://github.com/opensearch-project/OpenSearch/pull/877 — Added a broken link checker
> - https://github.com/opensearch-project/OpenSearch/issues/894 — Fixed broken links across the OpenSearch-Project
> - https://github.com/opensearch-project/opensearch-build/pull/17  — Updated labels across all OpenSearch repos
> - https://github.com/opensearch-project/opensearch-plugins/pull/70 — Updated plugin documentation
> 
> Vacha has worked tenaciously on non-glamorous tasks, because they raise the overall Engineering bar for the entire project. 

### Interest

Upon receiving at least 3 positive maintainer votes, and no vetoes, from other maintainers within a maximum period of 2 weeks, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository, which typically happens via e-mail obtained via DCO, but may also be a comment to a pull request, following a substantial contribution.

> > This is great work, Vacha! Would you be interested in becoming a [co-maintainer](https://github.com/opensearch-project/.github/blob/main/MAINTAINERS.md) on this project? I’d be honored to nominate you.

Author: You're absolutely right, the discussion has already happened and we should reflect that. I changed the language to ~ what you suggest: _This is great work! Based on your ongoing contributions to the project we have decided that we'd like to invite you to become a co-maintainer for this project. Please let us know whether you accept this nomination._

> > This is great work, Vacha! Would you be interested in becoming a [co-maintainer](https://github.com/opensearch-project/.github/blob/main/MAINTAINERS.md) on this project? I’d be honored to nominate you.

Author: For the second point I decided it's better to waste maintainers time than having a nearly qualified contributor be disappointed (the current version).

Individuals accept the nomination by replying, or commenting, for example _"Thank you! I would love to."_

### Public Confirmation

> Upon receiving 3 positive maintainer votes, and no vetoes, from other maintainers, and after having confirmed interest with the nominee, the maintainer opens a public issue in order to make the vote visible to the community. 

Reviewer 10: we discussed this one https://github.com/opensearch-project/.github/pull/59#discussion_r838051423

> _Content from the above nomination._
> 
> Here is my +1.
> 
> If you support this nomination, please add your +1 or a comment. 
> 
> If you do not support this nomination, we invite you to e-mail all/some of the current [project maintainers](https://github.com/opensearch-project/OpenSearch/blob/main/MAINTAINERS.md) in the next 2 weeks. We will keep your feedback private to this group and discuss offline. We will not publicly post the reasons for a decision not to add this co-maintainer, or publicize your name. 
> 
> Anyone, including maintainers, can publicly add +1s as requested in the nomination issue if they support the nomination or contact the maintainers in private if they oppose. The issue stays open for 2 weeks.

### Maintainer Decision

> If a positive decision has been reached, the nomination is accepted and the nominating maintainer can thus comment on the issue with _"maintainers has been added"_. If a decision cannot be reached for any reason, the nomination is rejected with _"a decision could not be reached"_ with no further explanation necessary.

Reviewer 10: One thing that is unclear to me here: 
it is easy to see the +1's after the 2 weeks, but how do we aggregate the opposing votes? is it a positive outcome when there are more +1's than opposing comments?

Reviewer 11: As written, it's not a straight up or down vote by majority.  If even one maintainer does not support a nomination, it's a veto, and the nomination doesn't go through.  If someone who's not a maintainer is not in favor, the maintainers can and should take it into consideration (and it may move them to veto) but by itself it wouldn't block the nomination.

Reviewer 10: I got a bit mixed up here. So there are two phases for becoming a maintainer once an existing maintainer chose to nominate someone:
1) nomination by at least 3 maintainers + 0 vetos (2 weeks duration)
2) public confirmation - community says what they think and maintainers take it into consideration (2 weeks duration)

If I understand correctly the Public Confirmation is a form of "Speak now or forever hold your peace" kinda thing. Which makes sense to me:)

Author: Correct. The 2 weeks is just a time limit to account for inactive maintainers. The first part, and any negative community feedback, happens in private because we are talking about actual humans, and if I have a strong reason why someone shouldn't be a maintainer I'm unlikely to want to share it publicly. The public confirmation is to make the process more transparent and to give an opportunity for non-maintainers to provide feedback.

Reviewer 13: So I wonder if two weeks is really necessary. I understand the desire to permit time for everyone to have a voice, but two weeks as the term seems long and arbitrary. I suspect some nominations might have a longer discussion, while others would seem obvious and everyone voting/opining quickly. Instead of a fixed length, why not suggest a minimum amount of time (72 hours?) and no maximum. (e.g. it stops when votes stops or discussion stops). 72 hours might be too fast, but it would at least guarantee that no one snuck a nomination in over the weekend.

Reviewer 10: What's the rush though? two weeks seems like a good amount of time to me. I propose we start with this and then see how it goes - maybe 72 hours will make more sense as we go along.

Author: I picked 2 weeks because people tend to take week long vacations and we want to give them an opportunity to respond. But we can definitely change it later.

Reviewer 5: Just adding some past experience on this one. As @ke4qqq mentioned 72 hours is a pretty [common voting period](https://www.apache.org/dev/pmc.html#send-the-board-a-notice-of-the-vote-to-add-someone). It's pretty much become the standard across all voting activities (nominations, smoke testing, publishing releases) on all Apache projects to avoid long tail requests/decisions.

Here's an example of what happens in practice if many folks are on vacation and a vote doesn't get traction:

```
Only XXXXXX replied, which suggests that other PMC members do not think [the nominee] is ready yet.

Let's close this vote, I will start a new one in the future if [ the nominee] is still active.

This vote has failed.
 ```

In this example a follow up email thread was started:

```
Hi,
In the previous vote for a new committer, I didn't miss the vote but indeed I preferred the option to wait more time for this contributor, while I didn't want to block if there were enough positive votes.
What is the appropriate way to do that, to show that we didn't miss the vote, should we vote +0?
```

With many followups that folks had either missed the vote or forgot to vote `+0`.

I prefer this practice instead of waiting 2 weeks of "I slept since then" or email pileups that cause amnesia. 

This isn't a deal breaker to further hold up this PR, but this PR has "only" been open 10 days. If we had this under vote we would have to wait 4 more days before merging; and you can see several folks getting antsy why this is still open. Food for thought.

### Adding the New Maintainer 

> The Repo [Admin](ADMINS.md) adjusts the new maintainer’s permissions accordingly. The nominating maintainer opens a pull request to add the new maintainer to MAINTAINERS.md, and closes the nomination issue when that is merged and the individual has been given maintainer-level permissions.

Reviewer 6: There is only one listed admin in this repo, is that accurate?

Author: I removed the link, it's misleading that it's just in this repo.

## Removing a Maintainer

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Author: We don't and I'd rather not. It feels like an unnecessary constraint, wdyt?

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Reviewer 2: I think @TheAlgo has a point. What is inactivity anyway? Can a maintainer constantly respond to issues but never merge anything?

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Author: I just don't think we need to micromanage what inactivity means for each repo. Different projects move forward at different pace. In my experience it's going to be something like "we haven't heard from X in a long time (TM), ...". If someone wants to propose words that improve just "inactivity" I'm open to it though.

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Reviewer 14: note also that people can be on vacation, off sick, temporarily on another project (e.g. if they're contributing for work), etc.

and for the vacation please note that not everyone has only two weeks per year (sick-days included), we in Switzerland have a minimum of four weeks, most have five or more (and sick days are of course not counted against your vacation) ;)
other european countries have even more (e.g. 5 weeks by law).
so it can well be that somebody is off for 1-2 months (accumulated vacation days, additional unpaid vacation, sabbatical, etc.).

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Reviewer 12: The implied assumption of `inactivity` is that person is not actively contributing to the project for a long period of time, we are talking about at least 8-12 months time period. I think this is reasonable but as @dblock says, it highly depends on project dynamics, I agree with that.

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Reviewer: I don't have a better word to replace inactivity at this moment but we can make it a bit verbose.

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Author: Please suggest specific changes if you believe they will improve the text.

> Removing a maintainer is a disruptive action that the community of maintainers should not undertake lightly. There are several reasons a maintainer might be removed from the project, such as inactivity, or taking actions that negatively impact the project.

Author: Feels like a judgement on length. I'll keep it the way it is now, please tell me if you feel strongly about it.

### Moving On

There are plenty of reasons that might cause someone to want to take a step back or even a hiatus from a project. Existing maintainers can choose to leave the project at any time, with or without reason, by making a pull request to move themselves to the "Emeritus" section of MAINTAINERS.md, and asking an admin to remove their permissions.

> ### Inactivity

Reviewer 2: - What happens if there are no active maintainers? Is that okay? 

- Is there a limit to 'intends to reengage'? E.g. Lots of people have great intents but actions are different story, so if someone is inactive but 'intends to reengage' over and over does it just kick the can?

- The role of 'repo owner' is used here but not defined.

Author: > * What happens if there are no active maintainers? Is that okay?

That would make the project unmaintained, and it's on the admin to find additional maintainers. I am open to additional text for that if you want to suggest something.

> * Is there a limit to 'intends to reengage'? E.g. Lots of people have great intents but actions are different story, so if someone is inactive but 'intends to reengage' over and over does it just kick the can?

I guess so. Is that a problem?

> * The role of 'repo owner' is used here but not defined.

That's my bad. Replaced by repo admin. Thanks.

Reviewer 7: How is an emeritus member reinstated? I see two options: 1) they are re-nominated by an existing maintainer, as though it's an entirely new addition. (Rationale: Merit doesn't expire, but it does get stale.) 2) They simply say "I want back in" and there's a nod from the rest of the team. The "nod" could be a vote, or it could be even more lightweight. FWIW, at Apache, an emeritus member returning requires a vote, but this is generally understood to be as lightweight as possible, and you don't say "no" without cause.

Author: Let's do a fast follow on emeritus, https://github.com/opensearch-project/.github/issues/61

Reviewer 2: > That would make the project unmaintained, and it's on the admin to find additional maintainers. I am open to additional text for that if you want to suggest something.

Suggestion:

_If the repo is left without any maintainers, either by maintainer inactivity or moving on, the repo is considered unmaintained. The repo admin will seek out new maintainers and note the maintenance status in the repo README file._

Author: > > That would make the project unmaintained, and it's on the admin to find additional maintainers. I am open to additional text for that if you want to suggest something.
> 
> Suggestion:
> 
> _If the repo is left without any maintainers, either by maintainer inactivity or moving on, the repo is considered unmaintained. The repo admin will seek out new maintainers and note the maintenance status in the repo README file._

I like these words, but isn't it the admin responsibility and should be added to ADMINS.md?

Author: Thinking about it more, wdyt of https://github.com/opensearch-project/.github/pull/59/commits/8b3890381b75854efdbe5623b6f5e3348fdb24b6?

Reviewer 2: @dblock 'If the inactive maintainer responds and intends to reengage, no action is required' -> If the inactive maintainer responds and reengages, no action is required

Author: Spot on, https://github.com/opensearch-project/.github/pull/59/commits/7fd6c112d5dac382bf60fadd1b359aeff465eea2.

If the maintainers on the project notice that another maintainer is no longer an active or visible member of the community, the first step will be to tell the repo owner to contact the inactive maintainer to check-in. If the inactive maintainer responds and intends to reengage, no action is required. If the maintainer does not respond after a week, or indicates they no longer wish to be a maintainer, the repo owner can remove the inactive maintainer’s permissions and make a pull request to add them to the "Emeritus" section of the MAINTAINERS.md.

> ### Negative Impact on the Project

Reviewer 15: Would not meeting the responsibilities outlined in the `MAINTAINERS.md` document fall into this section and constitute basis for removal too?

A separate conversation is whether riding on the coattails of this initiative we want to also revisit that document to make sure it still represents the bar we hold our maintainers to

Reviewer 11: I think almost all of that would fall under not being active/visible.  Did you have a specific thing in mind, Dave?

It's part of the same doc, so we can always take a look at it :)

Author: In general, in open-source one cannot expect or ask for time/work/engagement from others, so I'd be careful writing anything about maintainer responsibilities being a job.

Author: I think those responsibilities are "responsibilities are tenets" rather than "responsibilities are a job". A maintainer cannot be expected, for example, to respond to an issue in 24 hours, but can be expected to be responsive to issues. Does this help? Am I thinking about it wrong? Either way, if you think there are words that would clarify this in the current change to MAINTAINERS.md, I'll gladly consider!

Reviewer 15: Makes sense. On second review of the maintainers file it phrases them with terms like "regularly" instead of being prescriptive, so I don't think any changes are needed. Thanks!

Reviewer 2: I would suggest changing 'Actions' to 'Actions (or inactions)' which covers some of my concerns about inactivity.

Author: I think I disagree. Inactions rarely have negative impact as meant by this paragraph (code of conduct violations, security risks). We talk a lot about inactivity above to avoid associating the two.

Reviewer 2: @dblock Oh there are lots of ways inactions by maintainers negatively impact the project. Off the top of my head:
- Security: Failing to merge a security fix
- CoC: Failing to flag violations

Reviewer 10: Are maintainers alone expected to flag CoC violations? This is a community wide responsibility and i feel like maintainers are not in this alone. Its the community + the **admins** that are responsible for this imo.

Author: Certainly not, but this is a MAINTAINERS doc. In the CoC we have language that talks about reporting these. Speaking of which, https://github.com/opensearch-project/.github/pull/69.

> If the maintainers on the project notice that another maintainer is no longer an active or visible member of the community, the first step will be to tell the repo owner to contact the inactive maintainer to check-in. If the inactive maintainer responds and intends to reengage, no action is required. If the maintainer does not respond after a week, or indicates they no longer wish to be a maintainer, the repo owner can remove the inactive maintainer’s permissions and make a pull request to add them to the "Emeritus" section of the MAINTAINERS.md.

Reviewer 2: Does "Emeritus" afford anything aside from a mention in maintainers?

Author: Let's do a fast follow on emeritus, https://github.com/opensearch-project/.github/issues/61

> Actions that negatively impact the project will be handled by the Admins, in coordination with other maintainers, in balance with the urgency of the issue. Examples would be [Code of Conduct](CODE_OF_CONDUCT.md) violations and security risks.

Reviewer 12: ```suggestion
Actions that negatively impact the project will be handled by the Admins, in coordination with other maintainers, in balance with the urgency of the issue. Examples would be [Code of Conduct](CODE_OF_CONDUCT.md) violations, deliberate harmful actions and security risks.
```

The sabotage is sadly on the rise these days, probably worth to emphasize it, wdyt?

Author: Good call, added.

Reviewer 10: There are many places in the doc regarding discussing things between maintainers - are we going to create a mailing list? perhaps a space in the community forum for maintainers + admins only?
If there is no issue relating to this I can create one - WDYT?

Author: Feel free to open an issue! I know @nknize is exploring different options to enable easier conversations between maintainers in "private".

> Actions that negatively impact the project will be handled by the Admins, in coordination with other maintainers, in balance with the urgency of the issue. Examples would be [Code of Conduct](CODE_OF_CONDUCT.md) violations, deliberate harmful actions, and security risks.

Reviewer 2: - nit: Admins is capitalized and the rest of the roles throughout the document aren't.

### Interest

Upon receiving at least three positive (+1) maintainer votes, and no vetoes

> (-1), from existing maintainers after a one week period, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository via private e-mail message.

Author: https://github.com/opensearch-project/.github/pull/59/commits/0e23acbeb5b2263ac13a336bf4d94e4f2ab29772

> (-1), from existing maintainers after a one week period, the nominating maintainer asks a potential nominee whether they might be interested in becoming a maintainer on the repository via private e-mail message.

Reviewer 4: Good catch @dblock!

### Inactivity

> Maintainer status never expires. If a maintainer becomes inactive for a time (usually several months), their account may be deactivated by the repo admin for security reasons. A maintainer can reach out to the repo admin to get their permissions reinstated.

Reviewer 8: We cannot de-activate GitHub accounts. 

Perhaps replace:
> their account may be deactivated by the repo admin for security reasons.

With:
> the repo admin may revoke access to the repository for security reasons.
